docs/subnets/introduction-to-vm.md

```diff
@@ -0,0 +1,335 @@
+# Introduction to Virtual Machines
+
+## Overview
+
+This is part of a series of tutorials for building a Virtual Machine (VM):
+
+- Introduction to Virtual Machines (this article)
+- [How to Build a Simple VM](./create-a-virtual-machine-vm.md)
+- [How to Build a Complex VM](./create-a-vm-blobvm.md)
+
+A Virtual Machine is a blueprint for a blockchain and a blockchain instantiates from a VM, very much similar to a class-object relationship. A Virtual Machine can define following: issuance of transactions, transaction types, block structure, block building algorithm, keeping pending transactions in the mempool, gossiping transactions to the connected nodes, etc.
+
+## Blocks and State
+
+Virtual Machines can be broken down into 2 components: blocks and state. The functionality provided by VMs is to:
+
+- Define the representation of a blockchain's state
+- Represent the operations in that state
+- Apply the operations in that state
+
+Each block in the blockchain contains a set of state transitions. Each block is applied in order from the blockchain's initial genesis block to its last accepted block to reach the latest state of the blockchain.
+## Blockchain
+
+A blockchain has 2 components: **Consensus Engine** and **Virtual Machine**. VMs mainly deal with the implementation related to the block's structure: building and parsing. The Consensus Engine helps in reaching consensus on the block issued by VM. Here is a brief overview:
+
+1. A validator node wants to update the blockchain's state
+2. It creates a block with state transition details in it
+3. It issues the block to the consensus engine
+4. Consensus engine gossips the block within the network to reach consensus on it
+5. Depending upon the consensus results, the engine can either accept or reject the block
+6. Every virtuous node on the network should have the same verdict for a particular block
+7. It depends on the VM implementation on what to do with the accepted or rejected block
+
+The consensus engine has been already provided by `AvalancheGo` on the Avalanche network. Developers can implement and customize their own VM.
+
+## How to Load a VM
+
+VMs are created as a module, whose binary is registered by a node running `AvalancheGo`, against the **vmID** (binary file name must be vmID). VMID is a user-defined string that is zero-extended to a 32-byte array and encoded in CB58.
```

Comment:
I guess this is true.. it's a 32 byte array that is cb58 encoded. It's not necessarily a user defined string although I suppose that's how we've used it. cc @StephenButtolph